internal/remediation/override.go

```diff
@@ -0,0 +1,310 @@
+					}
+				}
+				if count < bestCount {
+					// Found a new candidate.
+					bestCount = count
+					bestVK = ver.VersionKey
+					if bestCount == 0 { // stop if there are 0 vulns remaining
+						break
+					}
+				}
+			}
+
+			if bestCount < len(vulnerabilities) {
+				// Found a version that fixes some vulns.
+				newPatches = append(newPatches, overridePatch{
+					PackageKey:  vk.PackageKey,
+					OrigVersion: vk.Version,
+					NewVersion:  bestVK.Version,
+				})
+			}
+		}
+
+		if len(newPatches) == 0 {
+			break
+		}
+
+		// Patch and re-resolve manifest
+		newManif, err := patchManifest(newPatches, result.Manifest)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		result, err = resolution.Resolve(ctx, cl, newManif)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		result.FilterVulns(opts.MatchVuln)
+
+		// If the patch applies to a package that was already patched before, update the effective patch.
+		for _, p := range newPatches {
+			idx := slices.IndexFunc(effectivePatches, func(op overridePatch) bool { return op.PackageKey == p.PackageKey && op.NewVersion == p.OrigVersion })
+			if idx == -1 {
+				effectivePatches = append(effectivePatches, p)
+			} else {
+				effectivePatches[idx].NewVersion = p.NewVersion
+			}
+		}
+	}
+
+	if len(effectivePatches) == 0 {
+		return nil, nil, errOverrideImpossible
+	}
+
+	// Sort the patches for deterministic output.
+	slices.SortFunc(effectivePatches, func(a, b overridePatch) int {
+		if c := a.PackageKey.Compare(b.PackageKey); c != 0 {
+			return c
+		}
+
+		return a.Semver().Compare(a.OrigVersion, b.OrigVersion)
+	})
+
+	return result, effectivePatches, nil
+}
+
+// getVersionsGreater gets the known versions of a package that are greater than the given version, sorted in ascending order.
+func getVersionsGreater(ctx context.Context, cl client.DependencyClient, vk resolve.VersionKey) ([]resolve.Version, error) {
+	sys := vk.Semver()
+	// Get & sort all the valid versions of this package
+	// TODO: (Maven) skip unlisted versions and versions on other registries
+	versions, err := cl.Versions(ctx, vk.PackageKey)
+	if err != nil {
+		return nil, err
+	}
+
+	cmpFunc := func(a, b resolve.Version) int { return sys.Compare(a.Version, b.Version) }
+	slices.SortFunc(versions, cmpFunc)
+	// Find the index of the next higher version
+	offset, vkFound := slices.BinarySearchFunc(versions, resolve.Version{VersionKey: vk}, cmpFunc)
+	if vkFound { // if the given version somehow doesn't exist, offset will already be at the next higher version
+		offset++
+	}
+
+	return versions[offset:], nil
+}
+
+// patchManifest applies the overridePatches to the manifest in-memory. Returns a copy of the manifest that has been patched.
+func patchManifest(patches []overridePatch, m manifest.Manifest) (manifest.Manifest, error) {
+	if m.System() != resolve.Maven {
+		return manifest.Manifest{}, errors.New("unsupported ecosystem")
+	}
+
+	// TODO: may need special handling for the artifact's type and classifier
```

Comment:
what is this TODO about?

Reply:
Updated the comment.
Basically, this currently doesn't look at or set the `type` or `classifier` fields, so it fails to actually create overrides for packages that have them

Thinking about it now, this might have the possibility of creating an infinite loop - I'll fix this in another PR.

Reply:
sounds good